### PR TITLE
[gnn-survey]sparse graph attention

### PR DESCRIPTION
**Pull request by joshchang1112**
Add sparse version of graph attention network.

### CODE REVIEW

**Comment by joshchang1112** [comment]
Hi Phil, 

If my PR needs some modification or anything needs to update, just feel free to tell me if you are available:)
Thanks!

Best Regards, 
Josh

File: research/gnn-survey/README.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+```
+
+4. Check test results after training.
+
+
+## Training examples
+
+* Reproduce gat results on cora (83.% average test accuracy):
+
+`python train.py --model gcn --gpu 0 --epochs 500 --lr=0.005 --weight_decay 5e-4 --dropout 0.6
```
Comment (ppham27) [inline research/gnn-survey/README.md:44]: Put these in code blocks and use separate lines:

```
python train.py \
  --model gcn \
  --gpu 0 \
  --epochs 500 \
  --lr=0.005 \
  --weight_decay 5e-4 \
  --dropout 0.6 \ 
  --hidden_dim 8 \
  --num_heads 8 \
  --save_dir models/cora/gat \
  --normalize_adj=True \
  --sparse_features=True
```

For example

Comment (ppham27) [inline research/gnn-survey/README.md:44]: Should be gat?

File: research/gnn-survey/layers.py [hunk 2]
```diff
@@ -117,3 +117,73 @@ def call(self, inputs):
     attn = self.dropout(attn)
     output = tf.matmul(attn, x)
     return output
+
+
+class SpGraphAttnLayer(tf.keras.layers.Layer):
```
Comment (ppham27) [inline research/gnn-survey/layers.py:122]: Usually, we want to avoid abbreviations. SparseGraphAttnLayer would be a better name.

File: research/gnn-survey/layers.py [hunk 3]
```diff
@@ -117,3 +117,73 @@ def call(self, inputs):
...
+  """ Single sparse graph attention layer."""
+
+  def __init__(self, output_dim, dropout_rate, alpha=0.2, concat=True, **kwargs):
```
Comment (ppham27) [inline research/gnn-survey/layers.py:125]: I don't understand the `concat=True` argument. It's not documented and looks like it actually refers to the elu activation?

File: research/gnn-survey/README.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+## Training examples
+
+* Reproduce gat results on cora (83.% average test accuracy):
+
+`python train.py --model gcn --gpu 0 --epochs 500 --lr=0.005 --weight_decay 5e-4 --dropout 0.6 
+  --hidden_dim 8 --num_heads 8 --save_dir models/cora/gat --normalize_adj=True --sparse_features=True`
+
+* Reproduce gcn results on cora (81.6% average test accuracy):
+
+`python train.py --model gcn --gpu 0 --epochs 400 --lr 0.01 --weight_decay 5e-4 --dropout 0.5
```
Comment (ppham27) [inline research/gnn-survey/README.md:49]: The flag looks like it's `dropout_rate` not `dropout`?

**Comment by ppham27** [comment]
> Hi Phil,
> 
> If my PR needs some modification or anything needs to update, just feel free to tell me if you are available:)
> Thanks!
> 
> Best Regards,
> Josh

Hi Josh,

Sorry for the delay. I left some small comments.

File: research/gnn-survey/README.md (hunk 1 above)
Comment (joshchang1112) [reply research/gnn-survey/README.md:44]: ok!

Comment (joshchang1112) [reply research/gnn-survey/README.md:44]: yes, sorry for typo:(

File: research/gnn-survey/layers.py (hunk 2 above)
Comment (joshchang1112) [reply research/gnn-survey/layers.py:122]: OK!

File: research/gnn-survey/layers.py (hunk 3 above)
Comment (joshchang1112) [reply research/gnn-survey/layers.py:125]: It should be removed, thanks:)

File: research/gnn-survey/README.md (hunk 4 above)
Comment (joshchang1112) [reply research/gnn-survey/README.md:49]: yes, another typo:(

**Comment by joshchang1112** [comment]
Hi Phil, 

Thanks for your review, I just modified the typo.
